components/datetime/src/format.rs

```diff
@@ -0,0 +1,136 @@
+use crate::date::DateTimeType;
+use crate::error::DateTimeFormatError;
+use crate::fields::{self, FieldLength, FieldSymbol};
+use crate::pattern::{Pattern, PatternItem};
+use crate::provider::DateTimeDates;
+use icu_data_provider::structs;
+use std::fmt;
+
+/// `FormattedDateTime` is a intermediate structure which can be retrieved as
+/// an output from `DateTimeFormat`.
+///
+/// The structure contains all the information needed to display formatted value,
+/// and it will also contain additional methods allowing the user to introspect
+/// and even manipulate the formatted data.
+///
+/// # Examples
+///
+/// ```
+/// # use icu_locale::LanguageIdentifier;
+/// # use icu_datetime::{DateTimeFormat, DateTimeFormatOptions};
+/// # use icu_datetime::DummyDateTime;
+/// # use icu_data_provider::InvariantDataProvider;
+/// # let langid: LanguageIdentifier = "en".parse()
+/// #     .expect("Failed to parse a language identifier.");
+/// # let provider = InvariantDataProvider;
+/// # let options = DateTimeFormatOptions::default();
+/// let dtf = DateTimeFormat::try_new(langid, &provider, &options)
+///     .expect("Failed to create DateTimeFormat instance.");
+///
+/// let date_time = DummyDateTime::new(2020, 9, 1, 12, 34, 28, 0);
+///
+/// let formatted_date = dtf.format(&date_time);
+///
+/// let _ = format!("Date: {}", formatted_date);
+/// ```
+pub struct FormattedDateTime<'l, T>
+where
+    T: DateTimeType,
+{
+    pub(crate) pattern: &'l Pattern,
+    pub(crate) data: &'l structs::dates::gregory::DatesV1,
+    pub(crate) date_time: &'l T,
+}
+
+impl<'l, T> fmt::Display for FormattedDateTime<'l, T>
+where
+    T: DateTimeType,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write_pattern(self.pattern, self.data, self.date_time, f).map_err(|_| std::fmt::Error)
+    }
+}
+// Temporary formatting number with length.
+fn format_number(
+    result: &mut impl fmt::Write,
+    num: usize,
+    length: &FieldLength,
+) -> Result<(), std::fmt::Error> {
+    write!(result, "{:0>width$}", num, width = (*length as u8) as usize)
+}
+
+// Temporary simplified function to get the day of the week
+fn get_day_of_week(year: usize, month: usize, day: usize) -> usize {
+    let t = &[0, 3, 2, 5, 0, 3, 5, 1, 4, 6, 2, 4];
+    let year = if month < 2 { year - 1 } else { year };
+    (year + year / 4 - year / 100 + year / 400 + t[month] + day + 1) % 7
+}
+
+pub fn write_pattern<T>(
+    pattern: &crate::pattern::Pattern,
+    data: &structs::dates::gregory::DatesV1,
+    date_time: &T,
+    w: &mut impl fmt::Write,
+) -> Result<(), DateTimeFormatError>
+where
+    T: DateTimeType,
+{
+    for item in &pattern.0 {
+        match item {
+            PatternItem::Field(field) => match field.symbol {
+                FieldSymbol::Year(..) => format_number(w, date_time.year(), &field.length)?,
```

Comment:
We established earlier that `format_number` is well-defined only on FieldLength::One and FieldLength::TwoDigit.  However, you don't enforce anywhere that the year field is necessarily one of those two lengths.  The data provider could give you a pattern that has `yyyyy`, for example.

Reply:
Added an assertion

Reply:
But it's not an assertion. It's not a bug if the data provider gives you a pattern like this. If the user overlays data with custom overrides, who knows what we might get.

I would rather you match on field.length, run format_number if it is One or TwoDigit, and return an error otherwise.

Reply:
Or, alternatively, use the TwoDigit behavior if specified, and fall back to the One behavior for all other widths.